keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2346 @@
+- Some resource drivers might not have any unsuitable nodes, for example
+  because they modify containers and that works on all nodes at all
+  times. Forcing such drivers to set an allow list would cause unnecessary
+  work.
+
+In its state for the Pod the scheduler plugin must remember when it rejected a
+node because of UnsuitableNodes. That information will be used in Post-filter
+to deallocate resources.
+
+#### Post-filter
+
+This is called when no suitable node could be found. If the Pod depends on ResourceClaims with delayed
+allocation, then deallocating one or more of these ResourceClaims may make the
+Pod schedulable after allocating the resource elsewhere. Therefore each
+ResourceClaim with delayed allocation is checked whether all of the following
+conditions apply:
+- allocated
+- not currently in use
+- it was the reason why some node could not fit the Pod, as recorded earlier in
+  Filter
+
+One of the ResourceClaims satisfying these criteria is picked randomly and deallocation
+is requested by setting the Deallocate field. The scheduler then needs to wait
+for the resource driver to react to that change and deallocate the resource.
+
+This may make it possible to run the Pod
+elsewhere. If it still doesn't help, deallocation may continue with another
+ResourceClaim, if there is one. To prevent deallocating all resources merely
+because the scheduler happens to check quickly, the next deallocation will only
+requested when there is none currently pending.
+
+At the moment, the scheduler has no information that might enable it to
+prioritize which resource to deallocate first. Future extensions of this KEP
+might attempt to improve this.
+
+#### Pre-score
+
+This is passed a list of nodes that have passed filtering by the resource
+plugin and the other plugins. The PodScheduling.PotentialNodes field
+gets updated now if the field doesn't
+match the current list already. If no PodScheduling object exists yet,
+it gets created.
+
+When there is a scheduler extender configured for the claim, creating and
+updating PodScheduling objects gets skipped because the scheduler
+extender handles filtering.
+
+#### Reserve
+
+A node has been chosen for the Pod.
+
+If using delayed allocation and the resource has not been allocated yet,
+the SelectedNode field of the ResourceClaim
+gets set here and the scheduling attempt gets stopped for now. It will be
+retried when the ResourceClaim status changes.
+
+If all resources have been allocated already,
+the scheduler adds the Pod to the ReservedFor field of its ResourceClaims to ensure that
+no-one else gets to use those.
+
+If some resources are not allocated yet or reserving an allocated resource
+fails, the scheduling attempt needs to be aborted and retried at a later time
+or when the ResourceClaims change.
+
+#### Unreserve
+
+The scheduler removes the Pod from the ReservedFor field because it cannot be scheduled after
```

Comment:
Unreserve is called when Reserve fails, so it should be the opposite operation for Reserve. I don't you want to clear ReservedFor here.

Reply:
I do not quite follow. In Reserve, ReservedFor gets checked and set if necessary. That we combined it with allocation in one particular case (delayed allocation) is an optimization, it's still the plugin's job to ensure that it was done.

According to https://github.com/kubernetes/kubernetes/blob/609db7ed0b1f2839e414c17d29fe4d76edc994bd/pkg/scheduler/framework/interface.go#L449-L453, Unreserve then gets called when the pod cannot run after all, right?

That implies that it must release it's reservation of the claim, otherwise no other pod can use that claim either.

Reply:
Oops, nvm. I was thinking that Reserve would set ReservedFor even if it wasn't allocated yet. That is not the case, so ignore my comment.